### DISCUSSION TITLE
Add menu for searching symbols by type/kind instead of name

### DISCUSSION THREAD

**Discussion by EriKWDev** [helix-editor/helix#6037] [Ideas] [answered]
I find myself not using the "symbol picker" feature very much (space + s) since in the large projects I work on there are so many symbols and I rarely know what I want just by name, but one is so that we don't have to have stupid naming conventions for types of different kinds because editor tooling support for the type-system has been great.

The types of queries I'm more often more interested in is
- "What are all the constants in this project?"
- "What are all the structs in this file?"
- "Show me all the enums"

One way of achieving this (in terms of ui and rust specifically, not thinking too hard about other languages atm) would be to support better queries/filters within the symbol picker search. 

An easy win for Rust at least, would be to instead of just containing the symbol's name in the search, the prefix like "const", "fn", "struct", "enum" would be part of the searchable names.

For example, if you have a file containing the struct
```rust
struct MyStruct {
  field: f32
}

fn my_func(a: &mut MyStruct) {
    // ..
}
```
and you open the symbol picker (or perhaps a separate new dialog that does this behaviour instead), it would list "struct MyStruct", "fn my_func" instead of just the names "MyStruct", "my_func" in a huge list of things.

That way, I would be able to filter on all structs by searching "struct", and then be able to narrow down the search by names once I see the list of all the structs. Similairly for enums, consts, functions etc.

I don't want to have to use naming conventions like calling all my structs `st_` or `enum_` like I've seen in some C++ projects when we have a language with such a robust type system.

I don't know if this would be feasible to implement in a general enough way to support many languages, but would be cool for Rust!

**Comment by EriKWDev** [comment]
A completely different approach to solve my problem might be to just have a fuzzy-search picker like the `/`-search, but instead of only showing a single result at a time, instead return the entire results list from the query live in a drop down like the file picker menu. The line matching could then be shown in the dropdown. and context on the right side, similair to the global "space + /" menu once you've hit enter.

Searching '/ + struct ' and then spamming "n" is not as nice as if it was just a dropdown immediately showing all the results.

This way it would be language agnostic.

**Comment by sudormrfbin** [comment, marked as answer]
Tracked by #5446.

**Reply by EriKWDev** [reply]
One important distinction to note here is that a the issue suggests an indicator "F", "C" for each item, but that would not necessarily make it easier to make the query with a filter like "only search functions in this file" since searching "F" in the symbol picker (if this indicator is taken into account at all) would return many things that aren't functions that just contain "F" in the name

**Reply by pascalkuthe** [reply]
My plan was to make multiple columns of the picker separably filterable. This will also be required/useful for other pickers.

See https://github.com/helix-editor/helix/pull/4687#issuecomment-1364609308 and https://github.com/helix-editor/helix/pull/4687#issuecomment-1364601524

**Comment by EriKWDev** [comment]
Maybe I just want [ripgrep](https://github.com/BurntSushi/ripgrep) integration in helix xP

**Reply by pascalkuthe** [reply]
we already have a global search that uses ripgrep internally, because ripgrep is written in rust we can just use it directly instead of shelling out

**Reply by EriKWDev** [reply]
Didn't know the global search used ripgrep. That's neat!

I wish it was updating live like the file picker as i type in a list though but that seems to be worked on already! Thanks for maintaining this awesome editor I use every day :)

**Comment by EriKWDev** [comment]
Yup #196 sounds more in line to what I would like to have xP I don't care too much about the global search with a list of file-names though and would rather just have all the search results in the picker instead.
